pennylane/qaoa/cost.py

```diff
@@ -21,6 +21,174 @@
 from pennylane import qaoa
 
 
+########################
+# Hamiltonian components
+
+
+def bit_driver(wires, b):
+    r"""Returns the bit-driver cost Hamiltonian.
+
+    This Hamiltonian is defined as:
+
+    .. math:: H \ = \ (-1)^{b + 1} \displaystyle\sum_{i} Z_i
+
+    where :math:`Z_i` is the Pauli-Z operator acting on the
+    :math:`i`-th wire and :math:`b \ \in \ \{0, \ 1\}`. This Hamiltonian is often used when
+    constructing larger QAOA cost Hamiltonians.
+
+    Args:
+        wires (Iterable or Wires): The wires on which the Hamiltonian acts
+        b (int): Either :math:`0` or :math:`1`. Determines whether the Hamiltonian assigns
+                 lower energies to bitstrings with a majority of bits being :math:`0` or
+                 a majority of bits being :math:`1`, respectively.
+
+    Returns:
+        .Hamiltonian:
+
+    **Example**
+
+    >>> wires = range(3)
+    >>> hamiltonian = qaoa.bit_driver(wires, 1)
+    >>> print(hamiltonian)
+    (1.0) [Z0] + (1.0) [Z1] + (1.0) [Z2]
+    """
+    if b == 0:
+        coeffs = [-1 for _ in wires]
+    elif b == 1:
+        coeffs = [1 for _ in wires]
+    else:
+        raise ValueError("'b' must be either 0 or 1, got {}".format(b))
+
+    ops = [qml.PauliZ(w) for w in wires]
+    return qml.Hamiltonian(coeffs, ops)
+
+
+def edge_driver(graph, reward):
+    r"""Returns the edge-driver cost Hamiltonian.
+
+    Given some graph, :math:`G`, with each node representing a wire, and a binary
+    colouring (each node/wire is assigned either :math:`|0\rangle` or :math:`|1\rangle`), the edge driver
+    cost Hamiltonian will assign a lower energy to edges with endpoint colourings
+    supplied in ``reward``.
+
+    For instance, if ``reward`` is ``["11"]``, then edges
+    with both endpoints coloured as ``1`` will be assigned a lower energy, while
+    the other colourings  (``"00"``, ``"10"``, and ``"01"``) will be assigned a higher energy.
+
+    See usage details for more information.
+
+    Args:
+         graph (nx.Graph): The graph on which the Hamiltonian is defined
+         reward (list[str]): The list of two-bit bitstrings that are assigned a lower energy by the Hamiltonian
+
+    Returns:
+        .Hamiltonian:
+
+    **Example**
+
+    As can be seen in the following example, ``"11"``, ``"10"``, and ``"01"`` will be assigned a lower
+    energy than ``"00"``:
+
+    >>> graph = nx.Graph([(0, 1), (1, 2)])
+    >>> hamiltonian = qaoa.edge_driver(graph, ["11", "10", "01"])
+    >>> print(hamiltonian)
+    (0.25) [Z0 Z1] + (0.25) [Z0] + (0.25) [Z1] + (0.25) [Z1 Z2] + (0.25) [Z2]
```

Comment:
I'm still confused with the definition of this Hamiltonian, even after looking at the example.

1. What does it mean for an edge to be assigned an energy? _States_ (actually eigenstates of the Hamiltonian) can be associated with an energy, but the edges just denote a pair of qubits.

2. I don't think that the connection between the Hamiltonian and the reward is too clear. Maybe good to add in text below the code snippet what the expectation values are for some states? Something like
"For example, a quick calculation gives the following expectation values:
<000|H|000> = 1 (I think)
<110|H|110> = -0.25 (I think)
meaning, a colouring that assigns "00" to the first two qubits has higher energy than a colouring that assigns them with "11" "

Reply:
The usage details are helpful, but I think the explanation beforehand can be made clearer

Reply:
Hopefully it's more clear now!